repos/system_upgrade/el7toel8/actors/peseventsscanner/libraries/library.py

```diff
@@ -235,113 +264,107 @@ def parse_architectures(architectures):
-                    add_packages_to_tasks(current, installed_out_pkgs, 'to_keep')
+                    add_packages_to_tasks(current, installed_out_pkgs, Task.keep)
 
                     if event.action in ('Split', 'Merged'):
-                        # Uninstall those RHEL 7 pkgs that are no longer on RHEL 8
+                        # Remove those RHEL 7 pkgs that are no longer on RHEL 8
                         in_pkgs_without_out_pkgs = filter_out_out_pkgs(event.in_pkgs, event.out_pkgs)
-                        add_packages_to_tasks(current, in_pkgs_without_out_pkgs, 'to_remove')
+                        add_packages_to_tasks(current, in_pkgs_without_out_pkgs, Task.remove)
 
                 if event.action in ('Renamed', 'Replaced', 'Removed'):
-                    add_packages_to_tasks(current, event.in_pkgs, 'to_remove')
+                    add_packages_to_tasks(current, event.in_pkgs, Task.remove)
 
         do_not_remove = set()
-        for package in current['to_remove']:
-            if package in tasks['to_keep']:
+        for package in current[Task.remove]:
+            if package in tasks[Task.keep]:
                 api.current_logger().warning(
                     '{p} :: {r} to be kept / currently removed - removing package'.format(
-                        p=package, r=current['to_remove'][package]))
-                del tasks['to_keep'][package]
-            elif package in tasks['to_install']:
+                        p=package, r=current[Task.remove][package]))
+                del tasks[Task.keep][package]
+            elif package in tasks[Task.install]:
                 api.current_logger().warning(
                     '{p} :: {r} to be installed / currently removed - ignoring tasks'.format(
-                        p=package, r=current['to_remove'][package]))
-                del tasks['to_install'][package]
+                        p=package, r=current[Task.remove][package]))
+                del tasks[Task.install][package]
                 do_not_remove.add(package)
         for package in do_not_remove:
-            del current['to_remove'][package]
+            del current[Task.remove][package]
 
-        for package in current['to_install']:
-            if package in tasks['to_remove']:
+        for package in current[Task.install]:
+            if package in tasks[Task.remove]:
                 api.current_logger().warning(
                     '{p} :: {r} to be removed / currently installed - keeping package'.format(
-                        p=package, r=current['to_install'][package]))
-                current['to_keep'][package] = current['to_install'][package]
-                del tasks['to_remove'][package]
-                del current['to_install'][package]
+                        p=package, r=current[Task.install][package]))
+                current[Task.keep][package] = current[Task.install][package]
+                del tasks[Task.remove][package]
+                del current[Task.install][package]
```

Comment:
You're here modifying the container you're iterating over this is a bad thing and can lead to problems.
If you would copy the indices (keys I assume) I think this could work.

Line 341:
```python
for package in list(current[Task.install].keys()):
```

Reply:
Alright, buffered the deleted entries into a set as in the block above.